### PR TITLE
Delete unused functionality in fv3fit

### PR DESCRIPTION
**Pull request by mcgibbon**
Since moving from xr.Dataset batches to tf.data.Dataset, StackedBatches and other sequence functionality in fv3fit is no longer used. This PR deletes that code.

Significant internal changes:
- Deleted StackedBatches, check_empty, preserve_samples_per_batch, shuffled, XyArraySequence, XyMultiArraySequence, and ThreadedSequencePreloader

Requirement changes:
- Added requirements for plotly, tensorboard, and matplotlib to fv3fit, since these are used already in fv3fit.wandb

Resolves #<github issues> [JIRA-TAG]

(Delete this for the commit message)
You are encouraged to check the PR against the [Code Review Checklist](https://paper.dropbox.com/doc/Code-Review-Checklist--A4lKrs~xg7w5Gsb39N6JLNQoAg-IlsYffZgTwyKEylty7NhY).

### CODE REVIEW

File: external/fv3fit/fv3fit/_shared/packer.py [hunk 1]
```diff
@@ -178,151 +165,10 @@ def _fill_empty_coords(da):
     return da
 
 
-class ArrayPacker:
```
Comment (mcgibbon) [inline external/fv3fit/fv3fit/_shared/packer.py:181]: This is moved to the bptt code, which is the last place it is used.

File: external/fv3fit/fv3fit/keras/_models/convolutional.py [hunk 2]
```diff
@@ -26,13 +26,6 @@
 UNSTACKED_DIMS = ("x", "y", "z", "z_interface")
 
 
-def multiply_loss_by_factor(original_loss, factor):
```
Comment (mcgibbon) [inline external/fv3fit/fv3fit/keras/_models/convolutional.py:29]: Now uses dense-training-style loss.

File: external/fv3fit/fv3fit/keras/_models/precipitative.py [hunk 3]
```diff
@@ -72,79 +70,6 @@ def call(self, dQ2: tf.Tensor) -> tf.Tensor:
...
 
 
-def multiply_loss_by_factor(original_loss, factor):
-    def loss(y_true, y_pred):
-        return tf.math.scalar_mul(factor, original_loss(y_true, y_pred))
-
-    return loss
-
-
-def get_losses(
```
Comment (mcgibbon) [inline external/fv3fit/fv3fit/keras/_models/precipitative.py:82]: Now uses dense-training-style loss.

File: external/fv3fit/fv3fit/keras/_models/shared/utils.py [hunk 4]
```diff
@@ -1,41 +1,10 @@
-from fv3fit._shared.packer import ArrayPacker
 import tensorflow as tf
-from typing import List, Optional, Sequence, Type
+from typing import List, Sequence, Type
 from fv3fit.emulation.layers import StandardNormLayer, StandardDenormLayer, NormLayer
 
 import numpy as np
 
 
-def get_input_vector(
```
Comment (mcgibbon) [inline external/fv3fit/fv3fit/keras/_models/shared/utils.py:9]: Moved to bptt training, which is the only place it is used.

File: external/fv3fit/setup.py [hunk 5]
```diff
@@ -17,6 +17,10 @@
     "dacite>=1.6.0",
     "wandb[media]>=0.12.1",
     "pace-util>=0.7.0",
+    "matplotlib",
+    "plotly",
+    "tensorboard",
+    "pandas",
```
Comment (mcgibbon) [inline external/fv3fit/setup.py:23]: These were already required (imported in module), just not added to setup.py

File: external/fv3fit/tests/keras/test_convolutional_network.py [hunk 6]
```diff
@@ -58,28 +57,22 @@ def test_standard_input_gives_standard_output():
 def test_convolutional_network_build_standard_input_gives_standard_output():
     fv3fit.set_random_seed(0)
     nt, nx, ny, nz = 5, 12, 12, 15
-    ds = xr.Dataset(
```
Comment (mcgibbon) [inline external/fv3fit/tests/keras/test_convolutional_network.py:61]: Decoupled from no-longer-used code.

File: external/fv3fit/fv3fit/_shared/scaler.py [hunk 7]
```diff
@@ -226,29 +127,3 @@ def loads(b: str) -> NormalizeTransform:
             return scaler_cls.load(f)
 
     raise NotImplementedError(f"Cannot load {class_name} scaler")
-
-
-def get_scaler(
```
Comment (AnnaKwa) [inline external/fv3fit/fv3fit/_shared/scaler.py:231]: I don't see any changes to the fv3fit RF module in this PR, was it changed previously to remove its usage of this function?

File: external/fv3fit/fv3fit/_shared/stacking.py [hunk 8]
```diff
@@ -11,40 +9,6 @@
 Z_DIM_NAMES = ["z", "pfull"]
 
 
-class StackedBatches(Sequence[xr.Dataset]):
```
Comment (AnnaKwa) [inline external/fv3fit/fv3fit/_shared/stacking.py:14]: Note that this means we must now always use the `unstacked_dim: z` configuration for keras training, might want to give the team a heads up (though I think @spencerkclark probably changed his config to do this anyways for recent experiments?)

File: external/fv3fit/fv3fit/_shared/stacking.py [hunk 9]
```diff
@@ -70,68 +34,6 @@ def stack_non_vertical(ds: xr.Dataset) -> xr.Dataset:
...
-def check_empty(ds: xr.Dataset) -> xr.Dataset:
-    """
-    Check for an empty variables along a dimension in a dataset
-    """
-    if len(ds[SAMPLE_DIM_NAME]) == 0:
-        raise ValueError("Check for NaN fields in the training data.")
-    return ds
-
-
-def preserve_samples_per_batch(ds: xr.Dataset) -> xr.Dataset:
```
Comment (AnnaKwa) [inline external/fv3fit/fv3fit/_shared/stacking.py:82]: @spencerkclark will the loaders still do this for the multiple climate datasets when you specify subsampling and `unstacked_dims: z` for stacking during the batch loading?

File: external/fv3fit/fv3fit/keras/_models/shared/__init__.py [hunk 10]
```diff
@@ -10,8 +10,7 @@
 from .training_loop import TrainingLoopConfig, EpochResult
 from .callbacks import TrainingLoopLossHistory
 from .loss import LossConfig
-from .utils import get_input_vector, standard_denormalize, standard_normalize
-from .sequences import XyArraySequence, XyMultiArraySequence, ThreadedSequencePreLoader
```
Comment (AnnaKwa) [inline external/fv3fit/fv3fit/keras/_models/shared/__init__.py:14]: IIRC `XyMultiArraySequence` used to handle appending halos for the convolutional model, was this moved somewhere else in a previous PR?

Comment (AnnaKwa) [reply external/fv3fit/fv3fit/keras/_models/shared/__init__.py:14]: Ok, I see a `append_halos_tensor` in the convolutional tests so I assume this is where that is done now?

File: external/fv3fit/fv3fit/keras/_models/loss.py [hunk 11]
```diff
@@ -1,148 +0,0 @@
-from typing import Callable, Union, Mapping, MutableMapping
```
Comment (AnnaKwa) [inline external/fv3fit/fv3fit/keras/_models/loss.py:1]: It's nice to delete this, it was kind of confusing when both this and `_models/shared/loss.py` were around.

**Review by AnnaKwa — COMMENTED** [review]
Thanks for the big cleanup, I just had a few questions to confirm that previous PRs had taken care of moving functionality that had been deleted.

File: external/fv3fit/fv3fit/_shared/stacking.py (hunk 8 above)
Comment (mcgibbon) [reply external/fv3fit/fv3fit/_shared/stacking.py:14]: Yes, or `[tile, x, y, z]` for convolutional training. This PR won't change that, it's currently the case in master.

File: external/fv3fit/fv3fit/keras/_models/shared/__init__.py (hunk 10 above)
Comment (mcgibbon) [reply external/fv3fit/fv3fit/keras/_models/shared/__init__.py:14]: Yes, it's done in the get_Xy helper equivalent for the convolutional training function.

File: external/fv3fit/fv3fit/_shared/stacking.py (hunk 9 above)
Comment (spencerkclark) [reply external/fv3fit/fv3fit/_shared/stacking.py:82]: Thanks for checking -- in my latest runs I have stopped doing this, so I am good with removing it.

File: external/fv3fit/fv3fit/_shared/stacking.py (hunk 8 above)
Comment (spencerkclark) [reply external/fv3fit/fv3fit/_shared/stacking.py:14]: Yup my latest config takes this into account.

**Comment by mcgibbon** [comment]
Extra commit was checked by @AnnaKwa .

**Review by AnnaKwa — APPROVED** [review]

**Pull request merged**